### ISSUE TITLE
Using LoadHtml/LoadUrl for displaying errors breaks back button

### ISSUE THREAD

**Issue by mitchcapper** [cefsharp/CefSharp#2985]
After https://github.com/cefsharp/CefSharp/issues/2921 the best way to handle errors seemed to be using LoadHtml/LoadUrl the problem of this is that it creates an additional history entry that prevents the back button from properly functioning. The back button takes you to the same error site that re-triggers the error.  We could work around this by not showing the error the second time but that would still require hitting the back button twice.   

A temporary solution is to inject JS to call the history.go(-2) if on an error page but isn't great.

Per @amaitland suggestion created an upstream discussion thread about suggested best practice with some suggested options:
https://www.magpcss.org/ceforum/viewtopic.php?f=10&t=17250

**Comment by mitchcapper** [comment]
OK! So I missed the initial reply but he pointed towards some code that seems like the correct way to do this, the InterstitialPageDelegate.

I found a decent example of this in another project for showing an ssl error page:
https://github.com/crosswalk-project/crosswalk/pull/2985/files


We would need to add some low level binders for the InterstitialPageDelegate but I don't think that would be too difficult.  I have a bit of an issue at work currently but once I get a bit of free time will go about implementing. I don't think this is a rush, but it certainly seems like a nicer way to do proper intermediate pages.

**Comment by amaitland** [comment]
Using DevTools it's possible to navigate to a specific history entry. 

- https://chromedevtools.github.io/devtools-protocol/tot/Page/#method-getNavigationHistory
- https://chromedevtools.github.io/devtools-protocol/tot/Page/#method-navigateToHistoryEntry

Haven't tested yet, will when I get s change. Still a workaround, might be a slightly nicer one.

**Comment by amaitland** [comment]
I think there's a better workaround for displaying errors, initial testing looks promising. Via `DevTools` you can use `Page.SetDocumentContent` to display the error page without having to hack around with the back/forward history.

```c#
chromiumWebBrowser.LoadError += OnBrowserLoadError;

private async void OnBrowserLoadError(object sender, LoadErrorEventArgs e)
{
	//Actions that trigger a download will raise an aborted error.
	//Aborted is generally safe to ignore
	if (e.ErrorCode == CefErrorCode.Aborted)
	{
		return;
	}

	//using LoadHtml/LoadUrl creates an additional history entry that
	//prevents the back button from working correctly.
	//Use Devools Page.SetDocumentContent to change the content
	using (var client = browser.GetDevToolsClient())
	{
		var response = await client.Page.GetFrameTreeAsync();
		var frames = response.FrameTree;
		var mainFrame = frames.Frame;

		var errorHtml = string.Format("<html><body><h2>Failed to load URL {0} with error {1} ({2}).</h2></body></html>",
									  e.FailedUrl, e.ErrorText, e.ErrorCode);

		_ = client.Page.SetDocumentContentAsync(mainFrame.Id, errorHtml);

		//AddressChanged isn't called for failed Urls so we need to manually update the Url TextBox
		this.InvokeOnUiThreadIfRequired(() => urlTextBox.Text = e.FailedUrl);
	}
}
```

The `WinForms MinimalExample` was updated in https://github.com/cefsharp/CefSharp.MinimalExample/commit/3cc3e642e9f84f20a5dabb12d23d6c332f063804 for testing purposes.

**Comment by amaitland** [comment]
Two new `SetMainFrameDocumentContentAsync` extension methods have been added  to simplify setting the main frame content. Important to note that `LoadError` won't trigger a `Address Changed` notification, if you have an address bar then you'll need to manually update it with the `FailedUrl`.

```c#
private void OnLoadError(object sender, LoadErrorEventArgs args)
{
	//Aborted is generally safe to ignore
	//Actions like starting a download will trigger an Aborted error
	//which doesn't require any user action.
	if(args.ErrorCode == CefErrorCode.Aborted)
	{
		return;
	}

	var errorHtml = string.Format("<html><body><h2>Failed to load URL {0} with error {1} ({2}).</h2></body></html>",
									  args.FailedUrl, args.ErrorText, args.ErrorCode);

	_ = args.Browser.SetMainFrameDocumentContentAsync(errorHtml);

	//AddressChanged isn't called for failed Urls so we need to manually update the Url TextBox
	this.InvokeOnUiThreadIfRequired(() => urlTextBox.Text = args.FailedUrl);
}
```

The example projects have been updated.

**Comment by amaitland** [comment]
The new extension methods are available in version 97.1.11.

**Comment by amaitland** [comment]
CEF will hopefully be adding a function for displaying error pages.

https://bitbucket.org/chromiumembedded/cef/issues/3265/alloy-add-better-support-for-error-pages
